src/aurora/gff3writer.cpp

```diff
@@ -0,0 +1,610 @@
+	if (iter != _labels.end()) {
+		return static_cast<uint32>(std::distance(_labels.begin(), iter));
+	} else {
+		_labels.push_back(label);
+		return static_cast<uint32>(_labels.size() - 1);
+	}
+}
+
+GFF3WriterStructPtr GFF3WriterList::addStruct(const Common::UString &label) {
+	// Create the structure pointer.
+	GFF3WriterStructPtr strct(new GFF3WriterStruct(_parent, static_cast<uint32>(_parent->_structs.size()) - 1));
+
+	// Create a field index.
+	_strcts.push_back(_parent->_structs.size());
+	GFF3Writer::Field field;
+	field.type = GFF3Struct::kFieldTypeStruct;
+	field.uint32Value = static_cast<uint32>(_parent->_structs.size());
+
+	// Add the label.
+	field.labelIndex = _parent->addLabel(label);
+
+	// Insert the newly created struct into the struct vector.
+	_parent->_structs.push_back(strct);
+
+	// Insert the struct to the field vector.
+	_parent->_fields.push_back(field);
+
+	return strct;
+}
+
+size_t GFF3WriterList::getSize() const {
+	return _strcts.size();
+}
+
+GFF3WriterList::GFF3WriterList(GFF3Writer *parent) : _parent(parent) {
+}
+
+uint32 GFF3WriterStruct::getId() const {
+	return _id;
+}
+
+size_t GFF3WriterStruct::getFieldCount() const {
+	return _fieldIndices.size();
+}
+
+GFF3WriterStructPtr GFF3WriterStruct::addStruct(const Common::UString &label) {
+	// Create the structure pointer.
+	GFF3WriterStructPtr strct(new GFF3WriterStruct(_parent, static_cast<uint32>(_parent->_structs.size()) - 1));
+
+	// Create a field index.
+	_fieldIndices.push_back(_parent->_fields.size());
+	GFF3Writer::Field field;
+	field.type = GFF3Struct::kFieldTypeStruct;
+	field.uint32Value = static_cast<uint32>(_parent->_structs.size());
+
+	// Add the label.
+	field.labelIndex = _parent->addLabel(label);
+
+	// Insert the newly created struct into the struct vector.
+	_parent->_structs.push_back(strct);
+
+	// Insert the struct to the field vector.
+	_parent->_fields.push_back(field);
+	
+	return strct;
+}
+
+GFF3WriterListPtr GFF3WriterStruct::addList(const Common::UString &label) {
+	// Create the list pointer.
+	GFF3WriterListPtr strct(new GFF3WriterList(_parent));
+
+	// Create a field index.
+	_fieldIndices.push_back(_parent->_fields.size());
+	GFF3Writer::Field field;
+	field.type = GFF3Struct::kFieldTypeList;
+	field.uint32Value = static_cast<uint32>(_parent->_lists.size());
+
+	// Add the label.
+	field.labelIndex = _parent->addLabel(label);
+
+	// Insert the newly created list into the lists vector.
+	_parent->_lists.push_back(strct);
+
+	// Insert the list to the field vector.
+	_parent->_fields.push_back(field);
+
+	return strct;
+}
+
+void GFF3WriterStruct::addByte(const Common::UString &label, byte value) {
+	// Create a field index.
+	_fieldIndices.push_back(_parent->_fields.size());
+
+	// Create field.
+	GFF3Writer::Field field;
+	field.type = GFF3Struct::kFieldTypeByte;
+	field.labelIndex = _parent->addLabel(label);
+	field.uint32Value = value;
+	_parent->_fields.push_back(field);
```

Comment:
Lots of code duplication here in these methods. Creating the field index and adding the field, i.e. everything except setting the actual value, can be abstracted into a method.